File: linklist/single/linklist.c

#include <stdio.h>
#include <stdlib.h>

typedef struct Node_t
{

    int data;
    struct Node_t *next;
} Node;

void insert_node_tail(Node **root, int element)
{

    Node *ptr;
    Node *first = *root;
    ptr = (Node *)malloc(sizeof(Node));
    ptr->data = element;

    if (*root == NULL)
    {
        *root = ptr;
    }
    else
    {

        while ((first->next) != NULL)
            first = first->next;

        first->next = ptr;
    }
}

void insert_node_head(Node **root, int element)
{

    Node *ptr;
    ptr = (Node *)malloc(sizeof(Node));
    ptr->data = element;

    if (*root == NULL)
    {
        *root = ptr;
    }
    else
    {

        ptr->next = *root;
        *root = ptr;
    }
}

void delete_node(Node **root, int element)
{

    printf("Deleting ... %d \n", element);

    Node *prev = NULL;
    Node *cur = *root;
    int found = 0;

    /*search for the element */
    while (cur->next != NULL)
    {

        if (cur->data == element)
        {
            found = 1;
            break;
        }
        prev = cur;
        cur = cur->next;
    }

    if (found == 1)
    {

        if (prev == NULL)
        {
            /* first element */
            *root = cur->next;
            free(cur);
        }
        else
        {
            prev->next = cur->next;
            free(cur);
        }
    }
    else
    {
        printf("No elements found \n");
    }
}

void display_node(Node *root)
{

    while (root)
    {
        printf("%d \t", root->data);
        root = root->next;
    }

    printf("\n");
}

void reverse_display_node(Node *root)
{

    if (root)
    {
        reverse_display_node(root->next);
        printf("%d \t", root->data);
    }

    printf("\n");
}

void reverse(Node **root)
{

    printf("Reversing \n");

    Node *prev = NULL;
    Node *cur = *root;
    Node *next = NULL;

    while (cur != NULL)
    {

        next = cur->next;
        cur->next = prev;
        prev = cur;
        cur = next;
    }

    *root = prev;
}

void insert_sorted(Node **root, int element)
{

    Node *ptr;
    Node *first = *root;
    Node *prev = NULL;
    ptr = (Node *)malloc(sizeof(Node));
    ptr->data = element;
     ptr->next = NULL;

    printf ("Inserting sorted node %d \n",element);

    if (*root == NULL)
    {
        *root = ptr;
    }
    else
    {

        while (1)
        {
            if (element < first->data)
            {
                printf (" Putting in \n");
                if (prev == NULL)
                {
                    ptr->next = *root;
                    *root = ptr;
                    return;
                }
                else
                {

                    prev->next = ptr;
                    ptr->next = first;
                    return;
                }
            } else {

                printf ("To next \n");
                prev = first; 
                first = first->next;
                

                if (first == NULL) {
                    /* reached end */
                    prev->next = ptr;
                    ptr->next = NULL; 
                    return;
                }    
            }    
        }
    }
}

int main()
{

    Node *root1 = NULL;
    Node *root2 = NULL;
    Node *root3 = NULL;

    printf("Single Linklist \n");

    insert_node_tail(&root1, 5);
    insert_node_tail(&root1, 10);
    insert_node_tail(&root1, 15);
    insert_node_tail(&root1, 20);

    display_node(root1);

    printf("Reverse displaying \n");
    reverse_display_node(root1);

    reverse(&root1);
    display_node(root1);

    insert_node_head(&root2, 100);
    insert_node_head(&root2, 200);
    insert_node_head(&root2, 300);
    insert_node_head(&root2, 400);

    display_node(root2);

    delete_node(&root2, 200);
    display_node(root2);

    insert_sorted (&root3, 99);
    insert_sorted (&root3, 1);
    insert_sorted (&root3, 10);
    insert_sorted (&root3, 2);
    display_node(root3);
}

File: linklist/single/linklist.go
package main

import (
	"fmt"
)

type  List struct {

	 data interface{};
	 next *List

};

func (l *List)insert_node_tail (element int) (*List) {

	m := l
	s := new(List)
	s.data = element
	if l == nil {
		fmt.Println ("Nil pointer")
		
        return s
	} else {

		for {
            if m.next != nil {
               m = m.next
		    } else {
				break
			}
		} 

        m.next = s
		return l

	}	    
}

func (l *List)insert_node_head (element int) (*List) {

	
	s := new(List)
	s.data = element
	if l == nil {
		fmt.Println ("Nil pointer")
		
        return s
	} else {
		s.next=l
		return s
	}	    
}

func (l *List)display_list ()  {
    
	fmt.Println ("Displaying ...") 
	

	for {
      if l != nil {
		fmt.Println (l.data)  
        l = l.next
	   } else {
		   break
	   }
	} 	    
}

func (l *List)delete_node (element int)  (*List){
    
	fmt.Printf ("Deleteing ... %d \n", element) 

	m := l
	var prev *List = nil

	found := 0
	
    // check if the element exists
	for {
      if l != nil {
		
     
		if l.data == element {
			found = 1
			break
		}
		prev = l
        l = l.next

	   } else {
		   break
	   }
	} 
	
	if found == 1 {

		if prev == nil {
			/* deleting the first element */
			return l.next
		} else {
			prev.next = l.next
			return m
		}

	} else {

		fmt.Println ("No elements found")
		return m
	}
}

func (l *List) reverse_display () {

	if (l != nil) {
	    l.next.reverse_display()
		fmt.Println (l.data) 
	}	
		

}

func (l *List) reverse () (*List){

	var prev *List = nil
	var cur  *List = l
	var next *List = nil

	fmt.Println ("Reversing ....")

	for {

		if (cur == nil) {
             return prev
		} else {
             next = cur.next
			 cur.next = prev
			 prev = cur
			 cur = next
		}
	}	
		

}


func (l *List) insert_sorted (element int) (* List) {

    var prev *List = nil
	//var cur *List = l
	var root *List = l

	s := new (List)
    s.data = element
	s.next = nil

	fmt.Printf ("Inserting ....%d \n",element)

	if (l == nil) {		

		return s
	} else {

		for {

			
			if element < l.data.(int) {

				if prev == nil {
					/* First element */

					s.next = l
					return s

				} else {
                   
					prev.next = s
					s.next = l

					return root
				}
				

			} else {
				fmt.Printf ("TO next....\n")
				prev = l
				l = l.next

				if l == nil {
					/* reach last element */
					prev.next = s
					return root
				}
			} 

		}

	}

	
	
}

func main () {

	var root *List = nil

	root = root.insert_node_tail (5)
	root = root.insert_node_tail (10)
	root = root.insert_node_tail (15)
	root = root.insert_node_tail (20)
	root.display_list ()
	root = root.reverse()
	root.display_list ()

	fmt.Println ("Reverse display 1 ......")
	root.reverse_display()

	var root1 *List = nil
	root1 = root1.insert_node_head (100)
	root1 = root1.insert_node_head (200)
	root1 = root1.insert_node_head (300)
	root1 = root1.insert_node_head (400)

	root1.display_list ()

	fmt.Println ("Reverse display 2 ......")
	root1.reverse_display()

	root1.delete_node(200)
	root1.display_list ()


	root1.delete_node(900)
	root1.display_list ()

	fmt.Println ("Insert sorted ....")
	var root2 *List = nil
	root2 = root2.insert_sorted (10)
	root2 = root2.insert_sorted (1)
	root2 = root2.insert_sorted (99)
	root2 = root2.insert_sorted (25)
	root2 = root2.insert_sorted (7)

	root2.display_list ()

    
}